firestore.rules

```diff
@@ -15,62 +27,192 @@ service cloud.firestore {
       allow read: if true;
     }
 
+    function teacherOfContext() {
+      return request.auth.token.user_type == 'teacher'
+        && request.auth.token.class_hash == resource.data.context_id
+        // The context_id is scoped to the platform_id
+        && request.auth.token.platform_id == resource.data.platform_id;
+    }
+
+    // Note: this conditional is used in read rules, which do not require a
+    // matching context_id, so the context_id is not enforced here
+    function learnerOwner() {
+      return request.auth.token.user_type == 'learner'
+        && string(request.auth.token.platform_user_id) == string(resource.data.platform_user_id)
+        && request.auth.token.platform_id == resource.data.platform_id;
+    }
+
+    function notChangingIdentifyingProps() {
+      // Make sure the learner isn't changing the document's identifying properties
+      return string(request.resource.data.platform_user_id) == string(request.auth.token.platform_user_id)
+        && request.resource.data.platform_id == request.auth.token.platform_id
+        // If for some reason the learner got a hold of a document from outside of their
+        // current class don't let them update it
+        && request.resource.data.context_id == request.auth.token.class_hash
+        // Make sure the learner isn't changing the current document's id
+        && request.resource.data.context_id == resource.data.context_id;
+    }
+
+    // This is similar to learnerOwner except that we are looking at the request.resource
+    // and in this case we can enforce the context_id matches the auth info
+    function learnerOwnerCreate() {
+      return request.auth.token.user_type == 'learner'
+        && string(request.auth.token.platform_user_id) == string(request.resource.data.platform_user_id)
+        && request.auth.token.platform_id == request.resource.data.platform_id
```

Comment:
Yeah I think so. There is a limit to the function stack but I think it is 10 so one nested function call should be fine.

Reply:
Oh actually looking at my own comment, we can't re-use it because in one case we are looking at `resource.data` and the other we are looking at `request.resource.data`.  It is possible to pass arguments to functions, so maybe we can parameterize it. I'll give it a shot...